url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/methodology/Evolution.java

```diff
@@ -0,0 +1,237 @@
+				//new DistanceFraudProbability(new DoubleRelation(), new LogScale(2)),
+				new TerroristFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				//new TerroristFraudProbability(new DoubleRelation(), new LogScale(2)),
+				new TotalBitsExchanged(new IntegerRelation(), new NoScale<Integer>()),
+				//new TotalBitsExchanged(new BitsExchangedRelation(), new NoScale<Integer>()),
+				new SizeOfMessages(new SizeOfMessagesRelation(), new NoScale<Integer>()),
+				//new SizeOfMessages(new IntegerRelation(), new NoScale<Integer>()),
+				new CryptoCalls(new IntegerRelation(), new NoScale<Integer>()),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+				//new Memory(new LongRelation(), new KbitsScale()),
+				new FinalSlowPhase(new FinalSlowPhaseRelation(), new NoScale<Boolean>()),
+		};
+		
+		
+		ParetoFrontier[] frontiers = ParetoFrontier.computeAllParetoFrontiers(protocols, attributes);
+		
+		System.out.println("Saving on disk");
+		
+		History.saveInDiskTheFrontiers(frontiers, "evolution.obj");
+		
+		
+		History.printHistory(protocols, attributes, frontiers);
+		
+		//frontiers =  orderFrontierByBitsExchanged(frontiers, attributes);
+		
+		System.out.println("Printing latex table");
+		
+		History.printLatexTable(frontiers, 0, 1, "evolution_table.tex");
+
+		Gnuplot.createDatafile(frontiers[0].getProtocols(), frontiers[0].getFrontier(), "all");
+		//Gnuplot.createDatafile(frontiers[0].getProtocols(), frontiers[0].getFrontier(), "all_relation");
+		//Gnuplot.createDatafile(frontiers[0].getProtocols(), frontiers[0].getFrontier(), "all_without_relation");
+		
+	}
+	
+	private static ParetoFrontier[] orderFrontierByBitsExchanged(
+			ParetoFrontier[] frontiers, Attribute[] attributes) {
+		TreeMap<Integer, List<DBProtocol>> result = new TreeMap<>();
+		for (int i = 0; i < frontiers.length; i++) {
+			DBProtocol[] protocols = frontiers[i].getFrontier();
+			for (int j = 0; j < protocols.length; j++) {
+				int e = protocols[j].getTotalBitsExchangedDuringFastPhase();
+				//System.out.println("Adding protocol "+protocols[j].getIdentifier()+" with e = "+protocols[j].getTotalBitsExchangedDuringFastPhase());
+				if (!result.containsKey(e)){
+					result.put(e, new LinkedList<DBProtocol>());
+				}
+				result.get(e).add(protocols[j]);
+			}
+		}
+		ParetoFrontier[] result2 = new ParetoFrontier[result.size()];
+		int index = 0;
+		for (List<DBProtocol> protocols : result.values()){
+			DBProtocol[] pArray = new DBProtocol[protocols.size()];
+			int pos = 0;
+			for (DBProtocol p : protocols) pArray[pos++] = p;
+			//System.out.println("Adding "+pArray.length+" protocols with e = "+pArray[0].getTotalBitsExchangedDuringFastPhase());
+			result2[index++] = new ParetoFrontier(pArray, pArray, attributes, null);
+		}
+		return result2;
+	}
+
+	public static void main2(String[] args) throws IOException, ClassNotFoundException {
+		ObjectInputStream in = new ObjectInputStream(new FileInputStream(new File("evolution.obj")));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. Execution paths that do not contain closure statements were detected. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).